services/spar/test-integration/Test/Spar/APISpec.hs

```diff
@@ -90,24 +94,46 @@ spec = do
       context "known IdP" $ do
         it "responds with request" $ do
           env <- ask
-          (_, _, cs . UUID.toText . fromIdPId -> idp) <- createTestIdP
+          let idp = cs . UUID.toText . fromIdPId $ env ^. teIdP . idpId
           get ((env ^. teSpar) . path ("/sso/initiate-login/" <> idp) . expect2xx)
             `shouldRespondWith` (\(responseBody -> Just (cs -> bdy)) -> all (`isInfixOf` bdy)
                                   [ "<html xml:lang=\"en\" xmlns=\"http://www.w3.org/1999/xhtml\">"
                                   , "<body onload=\"document.forms[0].submit()\">"
                                   , "<input name=\"SAMLRequest\" type=\"hidden\" "
                                   ])
 
-    describe "/sso/finalize-login" $ do  -- TODO: either use workingIdP or mock one locally.  the
-                                         -- latter is faster to run, but we need the former anyway,
-                                         -- so we might as well rely on that.
+    describe "/sso/finalize-login" $ do
       context "access denied" $ do
-        it "responds with 'forbidden'" $ do
-          pending
+        it "responds with a very peculiar 'forbidden' HTTP response" $ do
+          (idp, privcreds, authnreq) <- negotiateAuthnRequest
+          authnresp <- liftIO $ mkAuthnResponse privcreds idp authnreq False
+          sparresp <- submitAuthnResponse authnresp
+          liftIO $ do
+            -- import Text.XML
+            -- putStrLn $ unlines
+            --   [ cs . renderLBS def { rsPretty = True } . fromSignedAuthnResponse $ authnresp
+            --   , show sparresp
+            --   , maybe "Nothing" cs (responseBody sparresp)
+            --   ]
+            statusCode sparresp `shouldBe` 200
+            let bdy = maybe "" (cs @LBS @String) (responseBody sparresp)
+            bdy `shouldContain` "<?xml version=\"1.0\" encoding=\"UTF-8\"?>"
```

Comment:
Does the body only contain these four lines? If yes, then it's better to use `shouldBe`

Reply:
i specifically avoided that because there is a bit more text in there, and some of it really isn't crucial for the correctness, so i'd prefer to keep it like it is.